### ISSUE TITLE
(Spoilers) RPT Hospital Elevator Animation Bug

### ISSUE THREAD

**Issue by Polymega** [elishacloud/Silent-Hill-2-Enhancements#69]
@Bigmanjapan 
I have a theory... Do you remember how a value got changed after James enters the rowboat and then any subsequent times he goes into the rowboat his animation became corrupt? 

I believe that is happening with RPT when he stabs Maria. Check out the following videos (pay attention as he stabs): 

Correct: https://youtu.be/MlH49MyOvPA
Incorrect: https://youtu.be/raWuBdi0cOE

Why do I think this is an issue like the rowboat animation? Because I was able to replicate this on **both** PC and PCSX2. The first time I played this cutscene on both platforms the animation worked correctly. When I reloaded the save (for PCSX2 done by soft menu reset (L1 + R1 + L2 + R2 + Start + Select)) the animation became corrupt on the second attempt. 

Would you be able to find the address that isn't being reset for this cutscene?

**Comment by Polymega** [comment]
Another theory: 

- If Maria takes NO damage during hallway run: Bug occurs. (Because RPT never did stabbing animation prior to cutscene?)

- If Maria DOES take damage during hallway run: Animation works correctly.

**Comment by Bigmanjapan** [comment]
https://youtu.be/84XDBvrQCb4?t=41

So far I have observed three way it could go: 

1. RPT that is spawn outside of the hallway performs an attack animation, the one spawned in the hallway does a glitched attack animation (like in the video above);
2. RPT that is spawn outside of the hallway stays still, the one spawned in the hallway does a glitched attack animation;
3. RPT that is spawn outside of the hallway stays still, the one spawned in the hallway does a normal attack animation;

**Comment by Polymega** [comment]
Where he first attacks during chase will determine how delayed his cutscene attack animation is?! 

Welp... Now I'm even more confused.

**Comment by Bigmanjapan** [comment]
RPT hit animation definitely enables various functions which remain uncalled if there is no hit during the hallway chase.

There is an address that holds a timer value which starts counting down once RPT does a hit. Value is around 1800. If you manage to enable the cutscene and watch Maria's death moment while the value is still greater than 0, then the RPT cutscene animation will play.

If a player was too slow during the hallway and the value hit 0 or if you edit it to 0 manually, then RPT will remain still during the cutscene.

Fix is simple: change the address value to let's say 1800 once again (or any value that will be enough to overlap the RPT animation part of the cutscene) upon cutscene start.

sh2pc.exe+1B7A7C4 (01F7A7C4) 4 byte — cutscene index (Maria's death scene is 0x30)
sh2pc.exe+1B7AF4C (01F7AF4C) 4 byte — RPT animation timer

https://youtu.be/bF6JOSz-m0Y

There is only one issue: RPT starting animation cycle seems to be wrong, so have to look into that. At least what I have now ouputs the same result every time.

**Comment by Bigmanjapan** [comment]
Not an actual fix yet, just relevant info. 

Cutscenes timers:

sh2pc.exe+1B7A7B8 (01F7A7B8) float 
sh2pc.exe+1B7A7C8 (01F7A7C8) 4 bytes

sh2pc.exe+1B7AF4C (01F7AF4C) float — RPT attack delay timer

Delay values:

540 — hard action
960 — normal/easy
1800 — beginner

Right now it seems that the issue stems from RPT speed (walking, attacking etc) being increased for the hallway section of the game via some raggedy code that also make the moment of the stab happen sooner. 

In theory, that stab animation starting point should have a value that check against a cutscene timer value and starts the animation. At least this is how blur effects in the cutscenes work. Maybe something else gets checked against the timer and starts the animation. Either way, if the starting point value is found, then it should be possible to increase it to make animation start later.

**Comment by Bigmanjapan** [comment]
So it is possible to change starting point of the stab animation: https://youtu.be/vNBpdHCt1H8

**Comment by Polymega** [comment]
Great find! And can this adjustment be done regardless of what action difficulty is being used?

**Comment by Bigmanjapan** [comment]
My current fix is as raggedy as the code that is responsible for the issue since it involves corrupting hallway RPT. Works for all action difficulties.

https://youtu.be/E-QV2WQdCB8

sh2pc.exe+1BB1E00 (01FB1E00) 4 byte
sh2pc.exe+1BB1E00 (01FB1E0C) 4 byte

Set these two addresses to 0xFFFFFFFF value (one iteration, no constant write) right before the cutscene start to corrupt the hallway RPT and make the cutscene RPT play its animation properly.

Ideally, you would want to set the values using the cutscene index change as a condition but it might not work. If you try to set the values during the cutscene, then the cutscene RPT will corrupt too. Setting the values right on the cutscene index address value change might be fast enough not to corrupt cutscene RPT — I cannot check that, it has to be programmed and tried. If it doesn't work, then it should be possible to use James Z position coordinate to make the corruption. Since James has some inertion to his forward movement there is a certain point right before the cutscene trigger area, crossing which a player will inevitably start the cutscene. That coordinate point could be used as a corruption condition.

sh2pc.exe+1B7A7C4 (01F7A7C4) 4 bytes — cutscene index
The cutscene in question is 0x30.

James' position coordinates:

sh2pc.exe+1BB101C (01FB101C) float
sh2pc.exe+1BB1024 (01FB1024) float

**Comment by Bigmanjapan** [comment]
My current fix is as raggedy as the code that is responsible for the issue since it involves corrupting hallway RPT. Works for all action difficulties.

https://youtu.be/E-QV2WQdCB8

sh2pc.exe+1BB1E00 (01FB1E00) 4 byte
sh2pc.exe+1BB1E00 (01FB1E0C) 4 byte

Set these two addresses to 0xFFFFFFFF value (one iteration, no constant write) right before the cutscene start to corrupt the hallway RPT and make the cutscene RPT play its animation properly.

Ideally, you would want to set the values using the cutscene index change as a condition but it might not work. If you try to set the values during the cutscene, then the cutscene RPT will corrupt too. Setting the values right on the cutscene index address value change might be fast enough not to corrupt cutscene RPT — I cannot check that, it has to be programmed and tried. If it doesn't work, then it should be possible to use James Z position coordinate to make the corruption. Since James has some inertion to his forward movement there is a certain point right before the cutscene trigger area, crossing which a player will inevitably start the cutscene. That coordinate point could be used as a corruption condition.

sh2pc.exe+1B7A7C4 (01F7A7C4) 4 bytes — cutscene index
The cutscene in question is 0x30.

James' position coordinates:

sh2pc.exe+1BB101C (01FB101C) float
sh2pc.exe+1BB1024 (01FB1024) float

**Comment by Polymega** [comment]
> My current fix is as raggedy as the code that is responsible for the issue since it involves corrupting hallway RPT. Works for all action difficulties.

Sometimes the way to fix something that was made with spit and glue is to fix it with some more spit and glue. :) 

> Setting the values right on the cutscene index address value change might be fast enough not to corrupt cutscene RPT — I cannot check that, it has to be programmed and tried. 

I'll ask Aero if he would make us a script for testing down the line. 

Thank you, Bigmanjapan! This is a universal (across all platforms) bug that now has the potential to be resolved thanks to your research and development!

**Comment by Polymega** [comment]
@elishacloud 
Setting those two addresses to FFFFFFFF at cutscene initiation doesn't work as it renders Pyramid Head invisible. Using Bigmanjapan's other approach might be the way to go then: 

> Ideally, you would want to set the values using the cutscene index change as a condition but it might not work. If you try to set the values during the cutscene, then the cutscene RPT will corrupt too. Setting the values right on the cutscene index address value change might be fast enough not to corrupt cutscene RPT — I cannot check that, it has to be programmed and tried. If it doesn't work, then it should be possible to use James Z position coordinate to make the corruption.

**WHEN** 
Room ID 01FB7DAC (4 bytes) - 0000005B

**AND** 
James Position X/Z? 01FB101C (float) **IS GREATER THAN** 33185

**SET BOTH:**
sh2pc.exe+1BB1E00 (01FB1E00) 4 byte
sh2pc.exe+1BB1E0C (01FB1E0C) 4 byte

**TO:** 
FFFFFFFF (one iteration, no constant write)

**UPON ROOM ID END:**
Re-initialize this code each time the Room ID and James position conditions have been met again. (In case a player reloads a previous save point and re-initializes this area.)

**Comment by Polymega** [comment]
@Bigmanjapan 
Good news, your second recommended solution: 

> If it doesn't work, then it should be possible to use James Z position coordinate to make the corruption. Since James has some inertion to his forward movement there is a certain point right before the cutscene trigger area, crossing which a player will inevitably start the cutscene. That coordinate point could be used as a corruption condition.

Works consistently every time! Pyramid Head's attack is now always in sync to the rest of the cutscene animation. [Elisha implemented the following solution above.](https://github.com/elishacloud/Silent-Hill-2-Enhancements/issues/69#issuecomment-464523695)

However, there is one unintended consequence of corrupting the Pyramid Head that chases you: His footstep SFX goes on a continuous loop while the cutscene plays. 

https://youtu.be/orWj8aa5nyI

The speed/loudness/intensity of his footstep sounds varies each time you initiate the cutscene. Sometimes the footsteps go crazy-fast, other times they may not play at all. 

Is it possible to kill the footsteps during this part? If that can be resolved, then this fix should be good to check-in? Here is a test build with the current fix in place: [d3d8_b1.2.1315.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/2881393/d3d8_b1.2.1315.zip)

**Comment by Bigmanjapan** [comment]
Well, a rather quick and dirty solution would be setting RPT coordinates to 0 right before the cutscene in order to move him far way from the hospital maze location. Since footsteps' sound events are tied to his model position, it will be silent once he's positioned far away.

Bear in mind that the RPT coordinates addresses change depending on whether the hospital maze location is loaded via regular game progression (when savefiles are not used) or loaded via regular saving system / quickloading.

When loading is not used (addresses has to be confirmed one more time):

sh2pc.exe+1BB1B4C (01FB1B4) float — Z position
sh2pc.exe+1BB1B54 (01FB154)  float — X position

When loaded via quickloading:

sh2pc.exe+1BB1E18 (01FB1E18) float — Z position
sh2pc.exe+1BB1E20 (01FB1E20) float — X position

I will try to find another solution though.

**Comment by Polymega** [comment]
Actually, @Bigmanjapan , I tried out your solution of moving RPT far away and it seems to work great; the footsteps are gone. 

Unless you think there's an issues with it, we could go ahead and use this solution so we don't have to worry about it any more than we need to?

If we'd want to use this, here is what needs to be done: 

**Just like before:** 

**WHEN**
Room ID 01FB7DAC (4 bytes) - 0000005B

**AND**
James Position X/Z? 01FB101C (float) **IS GREATER THAN** 33185

**In addition to setting the previous addresses, ALSO SET BOTH:**
RPT Hospital X position (no load) sh2pc.exe+1BB1B54 (01FB154) float
RPT Hospital X position (game load) sh2pc.exe+1BB1E20 (01FB1E20) float

**TO:**
999999 (one iteration, no constant write)

**Comment by elishacloud** [comment]
@Polymega, Here is a test module.  See if this works:  [d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/2897590/d3d8.zip)

BTW: I think there is a typo in your addresses.  It should be:
 - RPT Hospital X position (no load) sh2pc.exe+1BB1B54 (01FB1**B**54) float

**Comment by Polymega** [comment]
> BTW: I think there is a typo in your addresses.

Good catch, thanks. 

Okay, so I **don't** think any of the following addresses affect RPT on a "fresh" game play (one where no game load has been used): 

```
RPT corruption addresses:
sh2pc.exe+1BB1E00 (01FB1E00) 4 byte
sh2pc.exe+1BB1E0C (01FB1E0C) 4 byte

RPT position addresses (no game load):
sh2pc.exe+1BB1B4C (01FB1B4) float — Z position
sh2pc.exe+1BB1B54 (01FB154) float — X position
```
Meaning, if a player doesn't load a game before getting to this point, the original bug still persists.

I originally didn't test this too much because I have to play the game fully from the start each time to test it. Even when cranking the internal game speed up it still takes a while each time to get back to this point.

**Comment by Polymega** [comment]
I will edit this as I find more, as I have to do a full, new playthrough each time. Once I get all the addresses, I can send you a revised code for this, Elisha: 

```
RPT corruption address (no game load):
sh2pc.exe+1BB293C (01FB293C) 4 byte
```

**Comment by Polymega** [comment]
Okay, I'm befuddled and must leave this to the professionals now. The addresses I found (one post above) keep changing every time I do a new game with no game load. 

In Cheat Engine, these addresses are always green in the search results, which means they're static addresses? 

In any case, on one game test, some of those addresses will corrupt RPT, and on other playthroughs they won't and the other addresses listed may corrupt him. After 5-6 playthroughs from the beginning to test this, I'm thoroughly confused now.

**Comment by Polymega** [comment]
Alright, I'm going to delete my most recent posts and consolidate everything to better explain: 

For this fix, as mentioned by Bigmanjapan, it turns out the game uses different sets of addresses depending on whether or not you loaded a game save file to get to this point or if you started a fresh, new game and made it all the way to this point from the beginning without loading back in. 

For all my previous play tests, I loaded right to this point for convenience and ease.

Our current fix iteration works every time for when you load a save file and get to this point. but not for "fresh" playthroughs. 

Every time I think I found the addresses to fix it for fresh playthroughs, different addresses are used for these values on my next playtest. The memory addresses are always the same to fix this when loading a save file to get to this point, but that doesn't seem to be the case for fresh playthroughs. 

Here are _some_ of the addresses I found that affect RPT's corruption and X position on fresh playthroughs. But again, these change each time. So some of the corruption addresses will corrupt RPT on one fresh playthrough, but other addresses will do so on another playthrough: 

```
RPT corruption addresses (no game load):
sh2pc.exe+1BB293C (01FB293C) 4 byte
sh2pc.exe+1BB3F90 (01FB3F90) 4 byte
sh2pc.exe+1BB3F9C (01FB3F9C) 4 byte
sh2pc.exe+1B7DF10 (01F7DF10) 4 byte

RPT position X (no game load):
sh2pc.exe+1BB4548 (01FB4548) float
```

And here is where our fix stands right now: 

```
WHEN
Room ID 01FB7DAC (4 bytes) - 0000005B

AND
James Position X/Z? 01FB101C (float) *IS GREATER THAN* 33185

SET:
RPT corruption pt. 1 (from game load) sh2pc.exe+1BB1E00 (01FB1E00) 4 byte
RPT corruption pt. 2 (from game load) sh2pc.exe+1BB1E0C (01FB1E0C) 4 byte
RPT corruption pt. 1 (no game load) sh2pc.exe+??????? (????????) 4 byte
RPT corruption pt. 2 (no game load) sh2pc.exe+??????? (????????) 4 byte

TO:
FFFFFFFF (one iteration, no constant write)

AND SET: 
RPT Hospital X position (from game load) sh2pc.exe+1BB1E20 (01FB1E20) float
RPT Hospital X position (no game load) sh2pc.exe+??????? (????????) float

TO: 
999999 (one iteration, no constant write)
```

**Comment by Bigmanjapan** [comment]
@Polymega 
Throughout the whole game HP values are assigned dynamically to a small region of addresses which in their turn could be traced via a single pointer address. I'll find the pointer a bit later.

I'll point out again that I don't really like this corruption solution since if you wait long enough after corrupting RPT (some internal RPT animation cycle?) the game will just crash. It only works because a player will step into a cutscene fast enough for the game not to crash.

**Comment by Polymega** [comment]
> I'll point out again that I don't really like this corruption solution since if you wait long enough after corrupting RPT (some internal RPT animation cycle?) the game will just crash. It only works because a player will step into a cutscene fast enough for the game not to crash.

I see. I wasn't aware of potential crash issues. If you're up to it, we could pause this current solution, to see if you find a more suitable fix?

**Comment by Bigmanjapan** [comment]
I tried another route of finding addresses that could put RPT to an unspawn state instead of corrupting him. 

sh2pc.exe+1B7AEA1 (01F7AEA1) byte — one of the RPT state addresses
Set it to 5 once upon `Room ID 01FB7DAC (4 bytes) - 0000005B` and `James Position X/Z? 01FB101C (float) *IS GREATER THAN* 33185`

This will suspend RPT without corrupting him (no game crashes). It will also make him silent, so there is no need in shifting his position to get rid of the footsteps' sounds.

**Comment by Polymega** [comment]
And it's consolidated down to a single address now? Beautiful, Bigmanjapan, _beautiful._

@elishacloud an update for the **HospitalChaseFix** mode. Things that are ~~strikethroughed~~ below will be dropped from the current iteration of this fix. Things that are **bold** will be added:

**WHEN**
Room ID 01FB7DAC (4 bytes) - 0000005B

**AND**
James Position X/Z? 01FB101C (float) is greater than 33185

**SET:**
**sh2pc.exe+1B7AEA1 (01F7AEA1) byte**
~~sh2pc.exe+1BB1E00 (01FB1E00) 4 byte~~
~~sh2pc.exe+1BB1E0C (01FB1E0C) 4 byte~~

**TO:**
**5 (one iteration, no constant write)**
~~FFFFFFFF (one iteration, no constant write)~~

**UPON ROOM ID END:**
Re-initialize this code each time the Room ID and James position conditions have been met again. (In case a player reloads a previous save point and re-initializes this area.)

**Comment by elishacloud** [comment]
Here is a test module with this fix implemented. Take a look at it and let me know if there are any issues with it. Test module: [d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/3024991/d3d8.zip)

_Note: this only works on v1.0._

**Comment by Polymega** [comment]
Played it both on a "no game load" run and a loaded save file run and it worked flawlessly both times. Great work you two!

**Comment by elishacloud** [comment]
I implemented a fix to get this working on all versions of SH2, however I am not completely sure the addresses on v1.1 and vDC are correct.  Try this one and see if it works on all versions:  [d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/3025818/d3d8.zip)

**Comment by Polymega** [comment]
Good news: This works on all versions. :)

**Comment by elishacloud** [comment]
Very good.  Implemented in check in [8795f76](https://github.com/elishacloud/Silent-Hill-2-Enhancements/commit/8795f765eac9fb1c7a8337c25a6dc3210b971ed4).

**Comment by Polymega** [comment]
Crazy that for 18+ years I always though RPT's bugged animation was just the way that was supposed to be. But nope, Bigmanjapan came to the rescue once more with his research and proposed solution! Thank you and Elisha for implementing the fix! 

An update has been released with this fix in it. See [v1.3.1340.0](https://github.com/elishacloud/Silent-Hill-2-Enhancements/releases/tag/v1.3.1340.0)
